000X-R-In-tree-pluggable-components.md

```diff
@@ -0,0 +1,132 @@
+
+## Background: business problem
+
+As of Dapr 1.10, we are shipping over 110 components across all of our building blocks, and the number is growing steadily with each release. This is great and it's the sign of a healthy OSS project with a growing community.
+
+However, this also brings some challenges. Dapr is written in Go, which as a technology incentivizes binaries to be statically compiled and self-contained. As of writing, there's no "plug-in" system for Go that works reliably across all operating systems. This means that the more components Dapr gains, the larger our binary becomes, and the more code is loaded in memory when `daprd` is started.
+
+The problems we've seen include, but are not limited to:
+
+* Amount of data loaded in memory.  
+  Dapr 1.10 has a virtual memory footprint of almost 900MB. Most of that is code pages that are not really a burden on the system, as they're offloaded easily by the kernel; nevertheless, the amount is growing with each release as we ship more code. A lot of that code is indeed due to components.  
+  More concerning, however, is when components instantiate memory when `daprd` is started, for example in `init` functions in Go packages. Often this memory is allocated on the heap, which also means that it causes more work for the garbage collector and has a negative impact on the performance of `daprd` overall.
+* Concerns with running on environments with limited resources.  
+  This is a corollary of the point above. Certain systems such as embedded ones or IoT devices have limited resources, including from storage and memory, so the increasing footprint of `daprd` is a concern in these scenarios. For example, we are aware of at least one large Dapr users who is re-building Dapr with a very limited subset of components to be able to fit it on embedded devices.
+* Effect of security vulnerabilities in third-party packages.  
+  Most of our components depend on third-party packages such as SDKs, and we import dozens of them in Dapr. Suppose that component X depended on a library for which a vulnerability was found. Users who do not leverage component X are likely not impacted by the vulnerability, but they are still running a binary that depends on a vulnerable package, possibly causing compliance conerns.
+* Ability to ship some components with outsized impact on performance.  
+  Already in the last months we've had components that were accepted in `dapr/components-contrib`, after being reviewed, and then we had to remove them because we noticed they had an unacceptable impact on the resources used by `daprd`. Sadly, our processes and tools don't allow us to catch these things easily until the component is actually registered in `daprd`.  
```

Comment:
It would be belter to anchor this time reference by specifically mentioning the last months of 1.10 release development. "In the last months" will be ambiguous 6 months from now.